### ISSUE TITLE
S-expression comments at the end of a pair causes a read error for a non-default readtable

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#1118]
I mentioned this on the mailing list, but after further investigation, I'm pretty sure this is a bug. The behavior is simple:

```racket
; this works
(read (open-input-string "(1 . 2 #;3)")) ; => '(1 . 2)

; this fails
(parameterize ([current-readtable (make-readtable (current-readtable))])
  (read (open-input-string "(1 . 2 #;3)"))) ; => read: illegal use of `.’
```

I am by no means qualified to read, much less understand, the C code, but my best guess is that it's an issue within `read.c`, specifically with the `skip_whitespace_comments` function. It branches on whether or not a readtable is present, and if it is, it unconditionally returns after skipping whitespace:

```c
// readtable.c, line 4124
if (table) {
  while ((ch = scheme_getc_special_ok(port), NOT_EOF_OR_SPECIAL(ch))) {
    if (!(readtable_kind(table, ch, params) & READTABLE_WHITESPACE))
break;
  }
  return ch;
} else {
  while ((ch = scheme_getc_special_ok(port), NOT_EOF_OR_SPECIAL(ch) && scheme_isspace(ch))) {}
}
```

A little bit below that, on line 4173, is some custom handling of s-expression comments. This will never be reached if a readtable is present because of the unconditional return. This is as far as I got, but I think it's probably on the right track: `skip_whitespace_comments` returns `#` instead of skipping the comment, so `effective_ch` is `#` within `read_list`, and a read error is raised.

**Comment by mflatt** [comment]
I agree that this is a bug, and it does look like the best fix is to make the rest of `skip_whitespace_comments` work with readtables. It looks like it should also deal with "specials" that turn into comments. I'll work on it.

**Comment by lexi-lambda** [comment]
Thanks for taking a look at it! I started tinkering with it a little bit myself, but I didn't get very far. I'm guessing it will need to be done with by looking up "effective character" mappings for readtables that remap characters to other readtables as well as handling readtable extensions that return special comments (my initial work-around attempt was to return a `make-special-comment`, before I looked at the C). I realized that I don't really know the C well enough at all to have any idea how to do that, though, especially given that implementing support for arbitrary special comments would involve lookahead.

Anyway, let me know if I can help out on this in any way. Either way, I appreciate your help in approaching this.
